### ISSUE TITLE
Rename IInitializeOptions to IPostConfigureOptions

### ISSUE THREAD

**Issue by danroth27** [aspnet/Options#191]
@divega @HaoK @ajcvickers 

Both `IConfigureOptions` and `IInitializeOptions` have similar semantics but they are called at different times. Based on the names it sounds like `IInitializeOptions` would happen before `IConfigureOptions`, but apparently the order is the opposite. Can we come up with a better name that makes the behavior more obvious?

**Comment by HaoK** [comment]
Yeah don't disagree, need a name that conveys this is a `PostConfigure` phase.

**Comment by danroth27** [comment]
What were the scenarios that motivated adding this additional stage? The scenarios might help determine the name. I believe you said one of the scenarios was validation?

**Comment by HaoK** [comment]
Its basically Fixup/Validation, but we didn't want to bite off the entire validation feature for options, so we targeted only fixup/normalization, auth uses it to do things to establish defaults, in your azure options for example, I use this to establish set the authority

**Comment by HaoK** [comment]
So if we want an uber lame name that's extremely clear, `service.PostConfigure<TOptions>()` does fit the description.

**Comment by danroth27** [comment]
@glennc

**Comment by divega** [comment]
Is this about triggering all the `IConfigureOptions` to run?

**Comment by HaoK** [comment]
867583d945f0297212ddc11d6863562d31cb251d

**Comment by divega** [comment]
A few of us talked about this with @davidfowl and he believes we really should revisit this decision.

**Comment by HaoK** [comment]
For post preview2 right?

**Comment by HaoK** [comment]
@divega can you summarize what we want to revisit, just the naming or its existence? :)

**Comment by divega** [comment]
@HaoK very briefly, @davidfowl thinks that the name `IPostConfigureOptions` is bad and wants us to think harder about it.

**Comment by divega** [comment]
@HaoK could you please summarize what `IPostConfigureOptions` is for?

**Comment by HaoK** [comment]
It should be thought of as a parallel set of Configure's that behave identically, but all run after Configure, in the order they were registered.  

The common scenario will be to calculate defaults from what's been already configuration.

**Comment by HaoK** [comment]
All that said PostConfigure is explicitly correct, any intent we give it is fairly arbitrary, its basically Configure2

**Comment by HaoK** [comment]
But something like [I]Calculate/Compute[Options] might imply running after Configure better than Initialize...

**Comment by divega** [comment]
@HaoK didn't this also allow us to avoid synchronization code somehow? Can you remind us how that works? And just to double-check, you can have multiple `IPostConfigureOptions`?

**Comment by HaoK** [comment]
Yeah so being able to register fixup code with options itself via `PostConfigure` means that fixup code is threadsafe and only run once which lets us avoid locking in auth.. 

`PostConfigure` behave exactly like `Configure`, you can add as many as you like, and they run in the order your add them.

Note these doesn't affect `IOptions`, they are consumed by `IOptionsFactory` which is used by `IOptionsSnapshot`

The factory simply runs all `IConfigureOptions`, followed by all `IPostConfigureOptions`

**Comment by HaoK** [comment]
We could make these also affect IOptions without too many sideeffects

**Comment by divega** [comment]
Thanks for clarifying. Then `IPostConfigureOptions` is only needed because we don't have another way to specify order? :disappointed: I assume no one has asked for `IPreConfigureOptions` yet :trollface: 

> We could make these also affect `IOptions` without too many side effects

Personally I think that would help make things a bit less confusing.

**Comment by HaoK** [comment]
I'll file an issue for whether IOptions should respect whatever we name this thing

**Comment by divega** [comment]
Given the description of the feature I don't think I can come up with a better name for it. Perhaps if we could bring order in without introducing a new type, that would be better, but previous attempts were not super nice, e.g. [having an Order property as an integer](https://github.com/aspnet/Options/blob/1.0.0-beta7/src/Microsoft.Framework.OptionsModel/IConfigureOptions.cs) lead to magic numbers scattered over the codebase.
